### PR TITLE
network: DHCP6 Allow to export DHCP6 DUID

### PR DESCRIPTION
**Pull request by ssahani**
We dont support the export of DHCP6 DUID. Let's add this. Some applications uses it.

```
"DHCPv6Client" : {
		"DUID" : [
			0,
			2,
			0,
			0,
			171,
			17,
			210,
			88,
			72,
			47,
			199,
			238,
			230,
			81
		]
	}


```

### CODE REVIEW

File: src/network/networkd-json.c [hunk 1]
```diff
@@ -1076,6 +1077,25 @@ static int dhcp6_client_pd_append_json(Link *link, JsonVariant **v) {
...
+        int r;
+
+        assert(link);
+        assert(v);
+
+        r = sd_network_link_get_dhcp6_client_duid_string(link->ifindex, &duid);
+        if (r < 0)
+                return r;
+
+        r = json_build(&w, JSON_BUILD_OBJECT(JSON_BUILD_PAIR_STRING("DUIDString", duid)));
```
Comment (yuwata) [inline src/network/networkd-json.c]: Please export DUID as a binary data, rather than as a string.

File: src/network/networkd-json.c [hunk 2]
```diff
@@ -1076,6 +1077,25 @@ static int dhcp6_client_pd_append_json(Link *link, JsonVariant **v) {
...
+        if (r < 0)
+                return r;
+
+        return json_variant_set_field_non_null(v, "DUID", w);
```
Comment (yuwata) [inline src/network/networkd-json.c]: Why nested??

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-json.c (hunk 2 above)
Comment (ssahani) [reply src/network/networkd-json.c]: Dropped

File: src/network/networkd-json.c (hunk 1 above)
Comment (ssahani) [reply src/network/networkd-json.c]: OK

**Comment by ssahani** [comment]
Thanks for the review. UPdated

File: src/network/networkd-json.c [hunk 3]
```diff
@@ -1076,6 +1077,25 @@ static int dhcp6_client_pd_append_json(Link *link, JsonVariant **v) {
...
+        assert(v);
+
+        if (!link->dhcp6_client || !sd_dhcp_duid_is_set(&link->dhcp6_client->duid))
+                return 0;
+
+        r = sd_dhcp_duid_get_raw(&link->dhcp6_client->duid, &data, &data_size);
+        if (r < 0)
+                return r;
+
+        return json_variant_merge_objectb(v, JSON_BUILD_OBJECT(JSON_BUILD_PAIR_HEX("DUID", data, data_size)));
```
Comment (yuwata) [inline src/network/networkd-json.c]: Please use `JSON_BUILD_PAIR_BYTE_ARRAY()`.

File: src/network/networkd-json.c [hunk 4]
```diff
@@ -1076,6 +1077,25 @@ static int dhcp6_client_pd_append_json(Link *link, JsonVariant **v) {
         return json_variant_set_field_non_null(v, "Prefixes", array);
 }
 
+static int dhcp6_client_duid_append_json(Link *link, JsonVariant **v) {
+        _cleanup_(json_variant_unrefp) JsonVariant *w = NULL;
```
Comment (yuwata) [inline src/network/networkd-json.c]: unused. Please drop.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-json.c [hunk 5]
```diff
@@ -20,6 +20,7 @@
 #include "networkd-route-util.h"
 #include "networkd-route.h"
 #include "networkd-routing-policy-rule.h"
+#include "sd-network.h"
```
Comment (yuwata) [inline src/network/networkd-json.c]: Please drop this.

File: src/network/networkd-json.c [hunk 6]
```diff
@@ -1076,6 +1076,24 @@ static int dhcp6_client_pd_append_json(Link *link, JsonVariant **v) {
...
 
+static int dhcp6_client_duid_append_json(Link *link, JsonVariant **v) {
+        const void *data;
+        size_t data_size;
+        int r;
+
+        assert(link);
+        assert(v);
+
+        if (!link->dhcp6_client || !sd_dhcp_duid_is_set(&link->dhcp6_client->duid))
```
Comment (yuwata) [inline src/network/networkd-json.c]: Please use `sd_dhcp6_client_get_duid()` instead of directly accessing the internal value.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly good.

**Review by yuwata — APPROVED** [review]

**Comment by ssahani** [comment]
Thanks for the review !

File: src/network/networkd-json.c [hunk 7]
```diff
@@ -1076,6 +1076,29 @@ static int dhcp6_client_pd_append_json(Link *link, JsonVariant **v) {
...
+        assert(v);
+
+        if (!link->dhcp6_client)
+                return 0;
+
+        r = sd_dhcp6_client_get_duid(link->dhcp6_client, &duid);
+        if (r < 0)
+                return 0;
+
+        r = sd_dhcp_duid_get_raw(&link->dhcp6_client->duid, &data, &data_size);
```
Comment (YHNdnzj) [inline src/network/networkd-json.c:1095]: Hmm, shouldn't this be `duid` (assigned by sd_dhcp6_client_get_duid above)?

Comment (yuwata) [reply src/network/networkd-json.c:1095]: Ouch. Follow-up commit is waiting in #31433.

**Pull request merged**